File: native-wrappers/ios/DatadogWrapper/Sources/DatadogWrapper/DatadogWrapper.swift
import Foundation
import DatadogCore
import DatadogInternal
import DatadogCrashReporting

@objc(DatadogWrapper)
public class DdSdkNativeWrapper: NSObject {

    // MARK: - Mapping helpers

    static func mapSite(_ site: String) -> DatadogSite {
        switch site.lowercased() {
        case "us3": return .us3
        case "us5": return .us5
        case "eu1": return .eu1
        case "ap1": return .ap1
        case "ap2": return .ap2
        case "us1_fed": return .us1_fed
        default: return .us1
        }
    }

    static func mapTrackingConsent(_ consent: String) -> TrackingConsent {
        switch consent.lowercased() {
        case "granted": return .granted
        case "not_granted": return .notGranted
        default: return .pending
        }
    }

    static func mapVerbosity(_ verbosity: String) -> CoreLoggerLevel {
        switch verbosity.lowercased() {
        case "debug": return .debug
        case "info": return .debug  // iOS has no info level, use debug
        case "warn": return .warn
        case "error": return .error
        default: return .error
        }
    }

    static func mapBatchSize(_ batchSize: String) -> Datadog.Configuration.BatchSize {
        switch batchSize.lowercased() {
        case "small": return .small
        case "large": return .large
        default: return .medium
        }
    }

    static func mapUploadFrequency(_ uploadFrequency: String) -> Datadog.Configuration.UploadFrequency {
        switch uploadFrequency.lowercased() {
        case "frequent": return .frequent
        case "rare": return .rare
        default: return .average
        }
    }

    static func mapBatchProcessingLevel(_ level: String) -> Datadog.Configuration.BatchProcessingLevel {
        switch level.lowercased() {
        case "low": return .low
        case "high": return .high
        default: return .medium
        }
    }

    static func mapProxyConfiguration(_ dict: NSDictionary?) -> [AnyHashable: Any]? {
        guard let dict = dict as? [String: Any],
              let type = dict["type"] as? String,
              let address = dict["address"] as? String,
              let port = dict["port"] as? Int else {
            return nil
        }

        var proxyDict: [AnyHashable: Any] = [:]

        switch type.lowercased() {
        case "http", "https":
            proxyDict["HTTPEnable"] = 1
            proxyDict["HTTPProxy"] = address
            proxyDict["HTTPPort"] = port
            proxyDict["HTTPSEnable"] = 1
            proxyDict["HTTPSProxy"] = address
            proxyDict["HTTPSPort"] = port
        case "socks":
            proxyDict["SOCKSEnable"] = 1
            proxyDict["SOCKSProxy"] = address
            proxyDict["SOCKSPort"] = port
        default:
            return nil
        }

        if let username = dict["username"] as? String {
            proxyDict[kCFProxyUsernameKey as String] = username
        }
        if let password = dict["password"] as? String {
            proxyDict[kCFProxyPasswordKey as String] = password
        }

        return proxyDict
    }

    // MARK: - First-party hosts storage

    /// Parsed first-party hosts stored during initialization.
    /// DdRum reads this when enabling RUM to configure urlSessionTracking on iOS.
    static var firstPartyHosts: [String: Set<TracingHeaderType>]? = nil

    /// Parses a flat dictionary of host -> comma-separated header types
    /// into the format expected by the Datadog SDK.
    static func parseFirstPartyHosts(_ dictionary: NSDictionary) -> [String: Set<TracingHeaderType>]? {
        guard let dict = dictionary as? [String: String] else { return nil }

        var hosts: [String: Set<TracingHeaderType>] = [:]
        for (host, headerTypesStr) in dict {
            let types = Set(
                headerTypesStr.split(separator: ",")
                    .map { DdRum.mapTracingHeaderType(String($0).trimmingCharacters(in: .whitespaces)) }
            )
            if !types.isEmpty {
                hosts[host] = types
            }
        }
        return hosts.isEmpty ? nil : hosts
    }

    // MARK: - Initialization

    /// Initialize the Datadog SDK with configuration
    /// - Parameters:
    ///   - clientToken: Datadog client token for your application
    ///   - environment: Environment name (e.g., "prod", "staging")
    ///   - service: Service name for your application (nullable)
    ///   - site: Datadog site (e.g., "us1", "eu1", "us3", "us5", "ap1", "us1_fed")
    ///   - verbosity: SDK verbosity level ("debug", "info", "warn", "error")
    ///   - trackingConsent: Initial tracking consent ("granted", "not_granted", "pending")
    ///   - batchSize: Batch size for data uploads
    ///   - uploadFrequency: Upload frequency for data batches
    ///   - batchProcessingLevel: Batch processing level
    ///   - additionalConfiguration: Additional configuration dictionary (includes _dd.version, _dd.version_suffix, etc.)
    ///   - firstPartyHosts: Dictionary of host -> comma-separated header types (nullable)
    /// - Returns: Boolean indicating successful initialization
    @objc public static func initialize(
        clientToken: String,
        environment: String,
        service: String?,
        site: String,
        verbosity: String,
        trackingConsent: String,
        batchSize: String?,
        uploadFrequency: String?,
        batchProcessingLevel: String?,
        proxyConfiguration: NSDictionary?,
        firstPartyHosts firstPartyHostsDict: NSDictionary?,
        nativeCrashReportEnabled: Bool,
        additionalConfiguration: NSDictionary?
    ) -> Bool {
        var configuration = DatadogCore.Datadog.Configuration(
            clientToken: clientToken,
            env: environment,
            site: mapSite(site),
            service: service
        )

        if let batchSize = batchSize {
            configuration.batchSize = mapBatchSize(batchSize)
        }

        if let uploadFrequency = uploadFrequency {
            configuration.uploadFrequency = mapUploadFrequency(uploadFrequency)
        }

        if let batchProcessingLevel = batchProcessingLevel {
            configuration.batchProcessingLevel = mapBatchProcessingLevel(batchProcessingLevel)
        }

        DatadogCore.Datadog.verbosityLevel = mapVerbosity(verbosity)

        // Apply additional configuration
        if let additionalConfig = additionalConfiguration as? [String: Any] {
            configuration._internal_mutation {
                $0.additionalConfiguration = additionalConfig
            }
        }

        if let proxyConfig = mapProxyConfiguration(proxyConfiguration) {
            configuration.proxyConfiguration = proxyConfig
        }

        // Store first-party hosts for RUM configuration
        if let hostsDict = firstPartyHostsDict {
            firstPartyHosts = parseFirstPartyHosts(hostsDict)
        } else {
            firstPartyHosts = nil
        }

        // Initialize Datadog SDK
        let core = DatadogCore.Datadog.initialize(
            with: configuration,
            trackingConsent: mapTrackingConsent(trackingConsent)
        )

        let initialized = !(core is DatadogInternal.NOPDatadogCore)

        if initialized {
            // DdRum.enableRum() will read this and activate crashReporting
            DdSdkNativeWrapper.nativeCrashReportEnabled = nativeCrashReportEnabled
        }

        return initialized
    }

    // Stored for DdRum to enable CrashReporting after RUM.enable()
    static var nativeCrashReportEnabled: Bool = false

    // MARK: - Tracking Consent

    // Dependencies - injectable for testing
    private static var datadogCore: DatadogSdkProtocol = RealDatadogSdk()

    static func setDatadogCore(_ core: DatadogSdkProtocol) {
        datadogCore = core
    }

    // Reset to production dependencies (for test cleanup)
    static func resetDependencies() {
        datadogCore = RealDatadogSdk()
        firstPartyHosts = nil
    }

    @objc public static func setTrackingConsent(_ consent: String) {
        datadogCore.setTrackingConsent(mapTrackingConsent(consent))
    }

    @objc public static func addAttribute(_ key: String, value: Any) {
        datadogCore.addAttribute(forKey: key, value: value)
    }

    @objc public static func addAttributes(_ attributes: NSDictionary) {
        guard let dict = attributes as? [String: Any] else { return }
        datadogCore.addAttributes(dict)
    }

    @objc public static func removeAttribute(_ key: String) {
        datadogCore.removeAttribute(forKey: key)
    }

    @objc public static func removeAttributes(_ keys: NSArray) {
        guard let keyList = keys as? [String] else { return }
        datadogCore.removeAttributes(forKeys: keyList)
    }

    // MARK: - User Info

    @objc public static func setUserInfo(_ id: String, name: String?, email: String?, extraInfo: NSDictionary) {
        let extra = (extraInfo as? [String: Any]) ?? [:]
        datadogCore.setUserInfo(id: id, name: name, email: email, extraInfo: extra)
    }

    @objc public static func addUserExtraInfo(_ extraInfo: NSDictionary) {
        let extra = (extraInfo as? [String: Any]) ?? [:]
        datadogCore.addUserExtraInfo(extra)
    }

    @objc public static func clearUserInfo() {
        datadogCore.clearUserInfo()
    }

    // MARK: - Account Info

    @objc public static func setAccountInfo(_ id: String, name: String?, extraInfo: NSDictionary) {
        let extra = (extraInfo as? [String: Any]) ?? [:]
        datadogCore.setAccountInfo(id: id, name: name, extraInfo: extra)
    }

    @objc public static func addAccountExtraInfo(_ extraInfo: NSDictionary) {
        let extra = (extraInfo as? [String: Any]) ?? [:]
        datadogCore.addAccountExtraInfo(extra)
    }

    @objc public static func clearAccountInfo() {
        datadogCore.clearAccountInfo()
    }

}

File: native-wrappers/android/datadogwrapper/src/main/kotlin/com/datadog/wrapper/DatadogWrapper.kt
/*
 * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
 * This product includes software developed at Datadog (https://www.datadoghq.com/).
 * Copyright 2026-Present Datadog, Inc.
 */

package com.datadog.wrapper

import android.content.Context
import android.util.Log
import com.datadog.android.Datadog
import com.datadog.android.DatadogSite
import com.datadog.android._InternalProxy
import com.datadog.android.core.configuration.BatchProcessingLevel
import com.datadog.android.core.configuration.BatchSize
import com.datadog.android.core.configuration.Configuration
import com.datadog.android.core.configuration.UploadFrequency
import com.datadog.android.ndk.NdkCrashReports
import com.datadog.android.privacy.TrackingConsent
import com.datadog.android.rum.GlobalRumMonitor
import com.datadog.android.trace.TracingHeaderType
import java.net.InetSocketAddress
import java.util.Locale
import java.net.Proxy
import okhttp3.Authenticator
import okhttp3.Credentials
import okhttp3.Request
import okhttp3.Response
import okhttp3.Route

class DatadogWrapper {
    companion object {

        private const val TAG = "DatadogWrapper"
        private const val PROXY_AUTHORIZATION_REQUIRED_STATUS_CODE = 407

        // -- Mapping helpers --

        @JvmStatic
        fun mapSite(site: String): DatadogSite = when (site.lowercase(Locale.US)) {
            "us1" -> DatadogSite.US1
            "us3" -> DatadogSite.US3
            "us5" -> DatadogSite.US5
            "eu1" -> DatadogSite.EU1
            "ap1" -> DatadogSite.AP1
            "ap2" -> DatadogSite.AP2
            "us1_fed" -> DatadogSite.US1_FED
            else -> DatadogSite.US1
        }

        @JvmStatic
        fun mapTrackingConsent(consent: String): TrackingConsent = when (consent.lowercase(Locale.US)) {
            "granted" -> TrackingConsent.GRANTED
            "not_granted" -> TrackingConsent.NOT_GRANTED
            else -> TrackingConsent.PENDING
        }

        @JvmStatic
        fun mapVerbosity(verbosity: String): Int = when (verbosity.lowercase(Locale.US)) {
            "debug" -> Log.DEBUG
            "info" -> Log.INFO
            "warn" -> Log.WARN
            "error" -> Log.ERROR
            else -> Log.ERROR
        }

        @JvmStatic
        fun mapBatchSize(batchSize: String): BatchSize = when (batchSize.lowercase(Locale.US)) {
            "small" -> BatchSize.SMALL
            "large" -> BatchSize.LARGE
            else -> BatchSize.MEDIUM
        }

        @JvmStatic
        fun mapUploadFrequency(uploadFrequency: String): UploadFrequency = when (uploadFrequency.lowercase(Locale.US)) {
            "frequent" -> UploadFrequency.FREQUENT
            "rare" -> UploadFrequency.RARE
            else -> UploadFrequency.AVERAGE
        }

        @JvmStatic
        fun mapBatchProcessingLevel(level: String): BatchProcessingLevel = when (level.lowercase(Locale.US)) {
            "low" -> BatchProcessingLevel.LOW
            "high" -> BatchProcessingLevel.HIGH
            else -> BatchProcessingLevel.MEDIUM
        }

        @JvmStatic
        fun mapProxyConfiguration(config: Map<String, Any>?): Pair<Proxy, Authenticator?>? {
            if (config == null) return null

            val type = (config["type"] as? String)?.lowercase(Locale.US) ?: return null
            val address = config["address"] as? String ?: return null
            val port = (config["port"] as? Number)?.toInt() ?: return null

            val proxyType = when (type) {
                "http", "https" -> Proxy.Type.HTTP
                "socks" -> Proxy.Type.SOCKS
                else -> return null
            }

            val proxy = Proxy(proxyType, InetSocketAddress(address, port))

            val username = config["username"] as? String
            val password = config["password"] as? String

            val authenticator: Authenticator? = if (username != null && password != null) {
                @Suppress("DEPRECATION_ERROR")
                Authenticator { _: Route?, response: Response ->
                    val proxyAuthorization = response.code() == PROXY_AUTHORIZATION_REQUIRED_STATUS_CODE

                    if (!proxyAuthorization) {
                        Log.w(
                            TAG,
                            "Unexpected response code=${response.code()}" +
                                " received during proxy authentication request."
                        )
                        return@Authenticator null
                    }

                    val challenges = response.challenges()
                    for (challenge in challenges) {
                        val scheme = challenge.scheme()
                        if ("Basic".equals(scheme, ignoreCase = true) ||
                            "OkHttp-Preemptive".equals(scheme, ignoreCase = true)
                        ) {
                            val credential = Credentials.basic(
                                username,
                                password,
                                challenge.charset()
                            )
                            return@Authenticator response.request().newBuilder()
                                .header("Proxy-Authorization", credential)
                                .build()
                        }
                    }

                    Log.w(
                        TAG,
                        "No known challenges are satisfied during proxy authentication request."
                    )
                    null
                }
            } else {
                null
            }

            return Pair(proxy, authenticator)
        }

        // -- Initialization --

        @JvmStatic
        fun mapTracingHeaderType(type: String): TracingHeaderType = when (type.lowercase(Locale.US)) {
            "b3" -> TracingHeaderType.B3
            "b3multi" -> TracingHeaderType.B3MULTI
            "tracecontext" -> TracingHeaderType.TRACECONTEXT
            else -> TracingHeaderType.DATADOG
        }

        // Parses a flat dictionary of host -> comma-separated header types
        // into the format expected by the Datadog SDK.
        @JvmStatic
        fun parseFirstPartyHosts(hosts: Map<String, Any?>): Map<String, Set<TracingHeaderType>> {
            val result = mutableMapOf<String, Set<TracingHeaderType>>()
            for ((host, value) in hosts) {
                val headerTypesStr = value?.toString() ?: continue
                val types = headerTypesStr.split(",")
                    .map { it.trim() }
                    .filter { it.isNotEmpty() }
                    .map { mapTracingHeaderType(it) }
                    .toSet()
                if (types.isNotEmpty()) {
                    result[host] = types
                }
            }
            return result
        }

        @JvmStatic
        fun initialize(
            context: Context,
            clientToken: String,
            environment: String,
            service: String?,
            site: String = "us1",
            verbosity: String = "error",
            trackingConsent: String = "pending",
            batchSize: String? = null,
            uploadFrequency: String? = null,
            batchProcessingLevel: String? = null,
            proxyConfiguration: Map<String, Any>? = null,
            firstPartyHosts: Map<String, Any?>? = null,
            nativeCrashReportEnabled: Boolean = false,
            additionalConfiguration: Map<String, Any>? = null
        ): Boolean {
            return try {
                val builder = Configuration.Builder(
                    clientToken = clientToken,
                    env = environment,
                    service = service
                )
                    .useSite(mapSite(site))

                batchSize?.let {
                    builder.setBatchSize(mapBatchSize(it))
                }

                uploadFrequency?.let {
                    builder.setUploadFrequency(mapUploadFrequency(it))
                }

                batchProcessingLevel?.let {
                    builder.setBatchProcessingLevel(mapBatchProcessingLevel(it))
                }

                additionalConfiguration?.let {
                    builder.setAdditionalConfiguration(it)
                }

                proxyConfiguration?.let { proxyConfig ->
                    mapProxyConfiguration(proxyConfig)?.let { (proxy, authenticator) ->
                        builder.setProxy(proxy, authenticator)
                    }
                }

                if (additionalConfiguration?.get("_dd.needsClearTextHttp") == true) {
                    _InternalProxy.allowClearTextHttp(builder)
                }

                // Configure first-party hosts for distributed tracing
                if (firstPartyHosts != null) {
                    val hosts = parseFirstPartyHosts(firstPartyHosts)
                    if (hosts.isNotEmpty()) {
                        builder.setFirstPartyHostsWithHeaderType(hosts)
                    }
                }

                builder.setCrashReportsEnabled(nativeCrashReportEnabled)

                Datadog.initialize(context, builder.build(), mapTrackingConsent(trackingConsent))

                Datadog.setVerbosity(mapVerbosity(verbosity))

                if (nativeCrashReportEnabled) {
                    NdkCrashReports.enable()
                }

                true
            } catch (e: Exception) {
                e.printStackTrace()
                false
            }
        }

        // -- Tracking Consent --

        @JvmStatic
        fun setTrackingConsent(consent: String) {
            Datadog.setTrackingConsent(mapTrackingConsent(consent))
        }

        // -- Global Attributes --

        @JvmStatic
        fun addAttribute(key: String, value: Any?) {
            GlobalRumMonitor.get().addAttribute(key, value)
        }

        @JvmStatic
        fun addAttributes(attributes: Map<String, Any?>) {
            val monitor = GlobalRumMonitor.get()
            for ((key, value) in attributes) {
                monitor.addAttribute(key, value)
            }
        }

        @JvmStatic
        fun removeAttribute(key: String) {
            GlobalRumMonitor.get().removeAttribute(key)
        }

        @JvmStatic
        fun removeAttributes(keys: List<String>) {
            val monitor = GlobalRumMonitor.get()
            for (key in keys) {
                monitor.removeAttribute(key)
            }
        }

        // -- User Info --

        @JvmStatic
        fun setUserInfo(id: String, name: String?, email: String?, extraInfo: Map<String, Any?>) {
            Datadog.setUserInfo(id, name, email, extraInfo)
        }

        @JvmStatic
        fun addUserExtraInfo(extraInfo: Map<String, Any?>) {
            Datadog.addUserProperties(extraInfo)
        }

        @JvmStatic
        fun clearUserInfo() {
            Datadog.clearUserInfo()
        }

        // -- Account Info --

        @JvmStatic
        fun setAccountInfo(id: String, name: String?, extraInfo: Map<String, Any?>) {
            Datadog.setAccountInfo(id, name, extraInfo)
        }

        @JvmStatic
        fun addAccountExtraInfo(extraInfo: Map<String, Any?>) {
            Datadog.addAccountExtraInfo(extraInfo)
        }

        @JvmStatic
        fun clearAccountInfo() {
            Datadog.clearAccountInfo()
        }

        // -- Flush --

        /**
         * Force the SDK to upload all buffered data immediately.
         *
         * Note: dd-sdk-android exposes flush via the internal proxy. This
         * also shuts down the upload executors, so the SDK should not be
         * used after calling flush() in the same process — it is intended
         * for end-of-test or end-of-process draining.
         */
        @JvmStatic
        fun flush() {
            Datadog._internalProxy().flushAndShutdownExecutors()
        }
    }
}
